src/core/DataFile.cpp

```diff
@@ -326,6 +336,139 @@ bool DataFile::writeFile( const QString& filename )
+								"write to this file. Please make sure you have write-access to "
+								"the file and try again.").arg(fullName));
+		}
+		qWarning() << "ERROR: Failed to create bundle file!";
+		return false;
+	}
+
+	// Creates resources folder
+	if (QDir(resourcesDirName).exists())
+	{
+		qWarning() << "Warning: Resources dir already exists. Proceeding anyways.";
+	}
+	else
+	{
+		if(!QDir().mkdir(resourcesDirName))
+		{
+			qWarning() << "ERROR: Failed to create resources dir!";
+			return false;
+		}
+	}
+
+	// Files will be named by the counter number + file extension
+	// to avoid name conflicts
+	unsigned int resCounter = 1;
+
+	std::map<QString,std::vector<QString>>::const_iterator it = ELEMENTS_WITH_RESOURCES.begin();
+
+	// Copy resources and manipulate the DataFile to have local paths to them
+	while(it != ELEMENTS_WITH_RESOURCES.end())
+	{
+		qWarning() << "Looking for resources on: " << it->first;
+		QDomNodeList list = elementsByTagName(it->first);
+
+		// Go through all elements with the tagname from our map
+		for (int i = 0; !list.item(i).isNull(); ++i)
+		{
+			QDomElement el = list.item(i).toElement();
+
+			std::vector<QString>::const_iterator res = it->second.begin();
+
+			// Search for attributes that point to resources
+			while(res != it->second.end())
+			{
+				qWarning() << "Attribute: " << *res;
+				// If the element has that attribute
+				if(el.hasAttribute(*res))
+				{
+					// Get absolute path to resource
+					QString resPath = PathUtil::toAbsolute(el.attribute(*res));
+					// The new file name is the counter number + the file extension
+					QString finalFileName = QString::number(resCounter) + "." + resPath.section('.',-1);
+					// Final path is our resources dir + the new file name
+					QString finalPath = resourcesDirName + "/" + finalFileName;
+
+					qWarning() << "Original Resource: " << resPath;
+					qWarning() << "Resource Destination: " << finalPath;
+
+					// Copy resource file to the resources folder
+					if(!QFile::copy(resPath, finalPath))
+					{
+						qWarning("ERROR: Failed to copy resource");
+						return false;
+					}
+
+					// Update attribute path to point to the bundle file
+					QString newAtt = "local:resources/" + finalFileName;
+					el.setAttribute(*res, newAtt);
+
+					qWarning() << "New Path: " << newAtt;
+					// Increases file counter
+					++resCounter;
+				}
+				++res;
+			}
+		}
+		++it;
+	}
+
+	// Writes the final DataFile
+	if (fullName.section('.', -1) == "mmpz")
+	{
+		QString xml;
+		QTextStream ts(&xml);
+		write(ts);
+		outfile.write(qCompress(xml.toUtf8()));
+	}
+	else
+	{
+		QTextStream ts(&outfile);
+		write(ts);
+	}
+
+	outfile.close();
+
+	// Make sure the file has been written correctly
+	if (QFileInfo(outfile.fileName()).size() > 0)
+	{
+		// Remove current file if there's one
+		QFile::remove(fullName);
+		// Move temporary file to current file
+		QFile::rename(fullNameTemp, fullName);
+
+		return true;
+	}
+
+	return false;
```

Comment:
We should consider refactoring this out of the function since it's a near copy/paste of `DataFile::writeFile`.

Maybe the opening code could be in `DataFile::openFile`, this writing part could be `DataFile::writeFile` (with boolean parameter `write_backup`), `DataFile::checkFile` for the written correctly part.

Reply:
True, I started from the `writeFile` method and ended up reusing too much of it. I'll try to extract parts of it to avoid repetition!